client/client.go

```diff
@@ -1484,166 +1572,206 @@ func (c *Client) runHandleInvocation(msg *wamp.Invocation) {
 		msg.ArgumentsKw = kwargs
 	}
 
-	// Create a kill switch so that invocation can be canceled.
-	var cancel context.CancelFunc
-	var ctx context.Context
-	if timeout > 0 {
-		// The caller specified a timeout, in milliseconds.
-		ctx, cancel = context.WithTimeout(context.Background(),
-			time.Millisecond*time.Duration(timeout))
-	} else {
-		ctx, cancel = context.WithCancel(context.Background())
-	}
-	c.invHandlerKill[reqID] = cancel
-	c.activeInvHandlers.Add(1)
-
-	// If caller is accepting progressive results, create map entry to allow
-	// progress to be sent.
-	if progResOK {
-		c.progGate[ctx] = reqID
+	cliInvocation := clientInvocation{
+		registration: msg.Registration,
+		request:      msg.Request,
 	}
-	c.sess.Unlock()
-
-	// Start a goroutine to run the user-defined invocation handler.
-	go func() {
-		defer cancel()
-
-		// Create channel to hold result.  Channel must be buffered.
-		// Otherwise, canceling the call will leak the goroutine that is
-		// blocked forever waiting to send the result to the channel.
-		resChan := make(chan InvokeResult, 1)
-		go func() {
-			// The Context is passed into the handler to tell the client
-			// application to stop whatever it is doing if it cares to pay
-			// attention.
-			resChan <- handler(ctx, msg)
-		}()
+	handlerQueue, queueExists := c.invHandlersQueues[cliInvocation]
+	ctx, _ := c.invHandlersCtxs[cliInvocation]
+	if !queueExists {
+		handlerQueue = make(chan *wamp.Invocation, 1)
+		c.invHandlersQueues[cliInvocation] = handlerQueue
 
-		// Remove the kill switch when done processing invocation.
-		defer func() {
-			c.sess.Lock()
-			delete(c.progGate, ctx)
-			delete(c.invHandlerKill, reqID)
-			c.sess.Unlock()
-			c.activeInvHandlers.Done()
-		}()
-
-		// Wait for the handler to finish or for the call be to canceled.
-		var result InvokeResult
-		select {
-		case result = <-resChan:
-			// If the handler returns InvocationCanceled, this means the
-			// handler canceled the call.
-			if result.Err == wamp.ErrCanceled {
-				c.log.Println("INVOCATION", reqID, "canceled by callee")
-			}
-		case <-c.Done():
-			c.log.Print("Client stopping, invocation handler canceled")
-			// Return without sending response to server.  This will also
-			// cancel the context.
-			return
-		case <-ctx.Done():
-			// Received an INTERRUPT message from the router.
-			// Note: handler is also just as likely to return on INTERRUPT.
-			result = InvokeResult{Err: wamp.ErrCanceled}
-			var reason string
-			if errors.Is(ctx.Err(), context.DeadlineExceeded) {
-				reason = "callee due to timeout"
-			} else {
-				reason = "router"
-			}
-			c.log.Println("INVOCATION", reqID, "canceled by", reason)
+		// Create a kill switch so that invocation can be canceled.
+		if timeout > 0 {
+			// The caller specified a timeout, in milliseconds.
+			ctx, cancel = context.WithTimeout(context.Background(),
```

Comment:
what does the timeout mean in this case? is it the timeout of a single invocation or the timeout for the whole request up until the final result?  if it's per invocation then there should be a new context for each invocation being sent to the handler

Reply:
Nope, this can not be a timeout for a single chunk, because again — there is no correlation between sending data and receiving results. 
In progressive calls the invocation starts with first call and lasts until receiving the final result. All intermediate calls (from client side) are still bound to the same invocation.

So in general I don't see how making  different contexts for every progressive call chunks can make sense... 
1st thought: may be update the previous one if it is provided?... Then it can be useful. Kind of pushing the deadline :)

Reply:
ok, so the timeout is for the whole complete invocation then? i.e. starts with the first invoke and must last until the final result?

Reply:
Yep!